server/src/main/java/org/elasticsearch/indices/recovery/PeerRecoveryTargetService.java

```diff
@@ -219,6 +220,7 @@ private void doRecovery(final long recoveryId, final StartRecoveryRequest preExi
         final RecoveryState recoveryState = recoveryTarget.state();
         final RecoveryState.Timer timer = recoveryState.getTimer();
         final IndexShard indexShard = recoveryTarget.indexShard();
+        final Releasable onCompletion = Releasables.wrap(recoveryTarget.disableRecoveryMonitor(), recoveryRef);
 
         final var failureHandler = ActionListener.notifyOnce(ActionListener.runBefore(ActionListener.noop().delegateResponse((l, e) -> {
```

Comment:
Kind of unrelated to your change, but I find the name `failureHandler` a bit confusing. Isn't this also called when the pre-recovery step completes successfully?

Reply:
Yeah I also don't 100% like this name, but nor can I think of a better one. The point is that this listener _only_ handles failures (i.e. it's a no-op on success, apart from releasing things that need releasing) and therefore needs further modification (i.e. a `.map()`) before it can be used as a proper listener.